File: apps/02-personal-md-files/backend/golang/src/services/llm/ollama_service.go
package llm

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

// OllamaService implements the LLMInterface for Ollama
type OllamaService struct {
	baseURL string
	model   string
	client  *http.Client
}

// NewOllamaService creates a new OllamaService
func NewOllamaService() (*OllamaService, error) {
	baseURL := os.Getenv("OLLAMA_HOST")
	if baseURL == "" {
		baseURL = "http://localhost:11434"
	}

	model := os.Getenv("OLLAMA_MODEL")
	if model == "" {
		model = "llama2"
	}

	client := &http.Client{
		Timeout: 30 * time.Second,
	}

	return &OllamaService{
		baseURL: baseURL,
		model:   model,
		client:  client,
	}, nil
}

// SendMessage sends a message to the Ollama API and returns the response
func (s *OllamaService) SendMessage(message string) (string, error) {
	requestBody := map[string]interface{}{
		"model":  s.model,
		"prompt": message,
		"stream": false,
	}

	jsonData, err := json.Marshal(requestBody)
	if err != nil {
		return "", fmt.Errorf("error marshaling request: %w", err)
	}

	req, err := http.NewRequest("POST", fmt.Sprintf("%s/api/generate", s.baseURL), bytes.NewBuffer(jsonData))
	if err != nil {
		return "", fmt.Errorf("error creating request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := s.client.Do(req)
	if err != nil {
		return "", fmt.Errorf("error making request to Ollama API: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("Ollama API error (status %d): %s", resp.StatusCode, string(body))
	}

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", fmt.Errorf("error decoding response: %w", err)
	}

	response, ok := result["response"].(string)
	if !ok {
		return "", fmt.Errorf("invalid response format from Ollama API")
	}

	return response, nil
}

// GetProvider returns the name of the LLM provider
func (s *OllamaService) GetProvider() string {
	return "Ollama (Local)"
}

// GetModel returns the model name used by the LLM provider
func (s *OllamaService) GetModel() string {
	return s.model
} 

File: apps/02-personal-md-files/backend/nodejs/src/services/llm/OllamaService.js
import axios from 'axios';
import LLMInterface from './LLMInterface.js';

class OllamaService extends LLMInterface {
  constructor() {
    super();
    this.baseUrl = process.env.OLLAMA_HOST || 'http://localhost:11434';
    this.model = process.env.OLLAMA_MODEL || 'llama2';
    this.axios = null;
    this.initializeAxios();
  }

  async initializeAxios() {
    this.axios = axios.create({
      baseURL: this.baseUrl,
      timeout: 30000,
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }

  async sendMessage(message) {
    try {
      const response = await this.axios.post('/api/generate', {
        model: this.model,
        prompt: message,
        stream: false,
      });

      if (!response.data?.response) {
        throw new Error('No response returned from Ollama API');
      }

      return response.data.response;
    } catch (error) {
      if (error.code === 'ECONNREFUSED') {
        throw new Error('Could not connect to Ollama server. Please ensure Ollama is running.');
      }
      throw new Error(`Error communicating with Ollama API: ${error.message}`);
    }
  }

  getProvider() {
    return 'Ollama (Local)';
  }

  getModel() {
    return this.model;
  }
}

export default OllamaService; 